fedimint-api/src/config.rs

```diff
@@ -170,6 +176,36 @@ impl FromStr for FederationId {
 }
 
 impl ClientConfig {
+    /// Returns the consensus hash for a given client config
+    pub fn consensus_hash(
+        &self,
+        module_config_gens: &ModuleGenRegistry,
+    ) -> anyhow::Result<sha256::Hash> {
+        let modules: BTreeMap<ModuleInstanceId, sha256::Hash> = self
+            .modules
+            .iter()
+            .map(|(module_instance_id, v)| {
+                let kind = v.kind();
+                Ok((
+                    *module_instance_id,
+                    module_config_gens
+                        .get(kind)
+                        .ok_or_else(|| format_err!("module config gen not found: {kind}"))?
+                        .hash_client_module(v.value().clone())?,
+                ))
+            })
+            .collect::<anyhow::Result<_>>()?;
+
+        let mut engine = HashEngine::default();
+        self.consensus_encode(&mut engine)?;
+        for (k, v) in modules.iter() {
+            k.consensus_encode(&mut engine)?;
+            v.consensus_encode(&mut engine)?;
+        }
+
+        Ok(sha256::Hash::from_engine(engine))
+    }
+
```

Comment:
That would be preferable but how would we access the `ModuleGenRegistry` in order to encode the submodules?

Reply:
Instead of the client module configs being `serde_json::Value`s internally, they could be already parsed an be a `Box<dyn IClientModuleConfig>` that allows encoding via `Encodable` as well as `serde`.

We already do this for module associated types (see `DynInput`, `DynOutput`, …), the only difference would be adding `serde` into the mix. We could even avoid that and just expose a from/to `serde_json::Value` fn (less complex than general serde support).

Reply:
I think the type would be `Box<dyn TypedClientModuleConfig>`, but if we went that route I'm not sure how we'd be able to derive things like `serde` and `Eq` on `ClientConfig`.  I thought the point of using `serde_json::Value` was because using `dyn` wasn't possible.

Reply:
I think `serde_json::Value` was just more convenient. I'll take a stab at it.

Reply:
If it works, perhaps we can do the same for `ServerConsensusConfig` modules